src/hotspot/share/runtime/thread.cpp

```diff
@@ -1585,157 +1581,109 @@ void JavaThread::remove_monitor_chunk(MonitorChunk* chunk) {
   }
 }
 
+void JavaThread::handle_special_runtime_exit_condition() {
+  if (is_obj_deopt_suspend()) {
+    frame_anchor()->make_walkable(this);
+    wait_for_object_deoptimization();
+  }
+  JFR_ONLY(SUSPEND_THREAD_CONDITIONAL(this);)
+}
+
 
 // Asynchronous exceptions support
 //
-void JavaThread::check_and_handle_async_exceptions() {
-  if (has_last_Java_frame() && has_async_exception_condition()) {
-    // If we are at a polling page safepoint (not a poll return)
-    // then we must defer async exception because live registers
-    // will be clobbered by the exception path. Poll return is
-    // ok because the call we are returning from already collides
-    // with exception handling registers and so there is no issue.
-    // (The exception handling path kills call result registers but
-    //  this is ok since the exception kills the result anyway).
-
-    if (is_at_poll_safepoint()) {
-      // if the code we are returning to has deoptimized we must defer
-      // the exception otherwise live registers get clobbered on the
-      // exception path before deoptimization is able to retrieve them.
-      //
-      RegisterMap map(this, false);
-      frame caller_fr = last_frame().sender(&map);
-      assert(caller_fr.is_compiled_frame(), "what?");
-      if (caller_fr.is_deoptimized_frame()) {
-        log_info(exceptions)("deferred async exception at compiled safepoint");
```

Comment:
Where is, or should be, this logging in the new code?

Reply:
It should be in ThreadSafepointState::handle_polling_page_exception(). Even in the old code that line should have been there since the previous place was just dead code (we don't call check_and_handle_async_exceptions() if is_at_poll_safepoint() is true). Added.